Fix offchain created from offchain

leoyvens opened the pull request:
This is blocking the update to the network subgraph, which now uses file data sources within file data sources.

This is more easily reviewed commit by commit. The first four commits are strictly refactors.

lutter commented on core/src/subgraph/context/instance.rs:
```diff
@@ -89,34 +90,31 @@ where
-            let host = this.new_host(logger.cheap_clone(), ds, module_bytes)?;
-            this.hosts.push(Arc::new(host));
+            let Some(host) = this.new_host(logger.cheap_clone(), ds)? else { continue };
+            this.hosts.push(host);
```
I think this would be cleaner as a `if let Some(..)` so we can avoid the `continue`

leoyvens replied:
This code changes in a further commit.

lutter commented on core/src/subgraph/context/instance/hosts.rs:
```diff
@@ -103,3 +111,90 @@ impl<C: Blockchain, T: RuntimeHostBuilder<C>> Hosts<C, T> {
 }
+
+/// Note that unlike `OnchainHosts`, this does not maintain the order of insertion. Ultimately, the
+/// processing order should not matter because as each offchain ds has its own causality region.
```
Small typo: `s/because as each/because each/`

lutter commented on core/src/subgraph/context/instance/mod.rs:
```diff
@@ -24,9 +24,11 @@ pub(super) struct SubgraphInstance<C: Blockchain, T: RuntimeHostBuilder<C>> {
-    /// `fn hosts_for_trigger` will return the hosts in the same order as their were inserted.
-    hosts: Hosts<C, T>,
+    /// Onchain hosts must be created in increasing order of block number. `fn hosts_for_trigger`
+    /// will return the onchain hosts in the same order as their were inserted.
```
`s/as their were/as they were/`

lutter commented on core/src/subgraph/context/instance/mod.rs:
```diff
@@ -127,28 +130,44 @@ where
-            Some(host)
-        })
+        // Check for duplicates and add the host.
+        match is_onchain {
```
Wouldn't a simple `if` be clearer here?

lutter commented on core/src/subgraph/context/instance/mod.rs:
```diff
@@ -127,28 +130,44 @@ where
+                        .last()
+                        .and_then(|h| h.creation_block_number())
+                        <= host.data_source().creation_block()
+                );
```
I am always a bit queasy with asserts since it's not that easy to spot panics in the logs; could we return an error here? That way, the subgraph will fail and the failure can tell use what exactly is wrong.

leoyvens commented on tests/tests/runner_tests.rs:
```diff
@@ -180,6 +180,63 @@ async fn data_source_revert() -> anyhow::Result<()> {
+    Ok(())
+}
+
+async fn data_source_long_revert() -> anyhow::Result<()> {
```
Yhea I'll squash merge

leoyvens commented:
@lutter thanks for the review! I've addressed it and removed the stray `dbg!`s I forgot to remove in https://github.com/graphprotocol/graph-node/pull/5083.
